enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/tx/edge/BatchingTxApplier.java

```diff
@@ -0,0 +1,161 @@
+import java.util.function.Supplier;
+
+import org.neo4j.kernel.impl.api.TransactionCommitProcess;
+import org.neo4j.kernel.impl.api.TransactionQueue;
+import org.neo4j.kernel.impl.api.TransactionToApply;
+import org.neo4j.kernel.impl.transaction.CommittedTransactionRepresentation;
+import org.neo4j.kernel.impl.transaction.log.TransactionIdStore;
+import org.neo4j.kernel.internal.DatabaseHealth;
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.kernel.monitoring.Monitors;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.neo4j.kernel.impl.transaction.tracing.CommitEvent.NULL;
+import static org.neo4j.storageengine.api.TransactionApplicationMode.EXTERNAL;
+
+/**
+ * Accepts transactions and queues them up for being applied in batches.
+ */
+public class BatchingTxApplier extends LifecycleAdapter implements Runnable
+{
+    private final int maxBatchSize;
+    private final Supplier<TransactionIdStore> txIdStoreSupplier;
+    private final Supplier<TransactionCommitProcess> commitProcessSupplier;
+    private final Supplier<DatabaseHealth> healthSupplier;
+
+    private final PullRequestMonitor monitor;
+    private final Log log;
+
+    private final ArrayBlockingQueue<CommittedTransactionRepresentation> txQueue;
+
+    private TransactionQueue txBatcher;
+
+    private volatile long lastQueuedTxId;
+    private volatile long lastAppliedTxId;
+
+    public BatchingTxApplier( int maxBatchSize, Supplier<TransactionIdStore> txIdStoreSupplier, Supplier<TransactionCommitProcess> commitProcessSupplier,
+            Supplier<DatabaseHealth> healthSupplier, Monitors monitors, LogProvider logProvider )
+    {
+        this.maxBatchSize = maxBatchSize;
+        this.txIdStoreSupplier = txIdStoreSupplier;
+        this.commitProcessSupplier = commitProcessSupplier;
+        this.healthSupplier = healthSupplier;
+
+        this.log = logProvider.getLog( getClass() );
+        this.monitor = monitors.newMonitor( PullRequestMonitor.class );
+
+        this.txQueue = new ArrayBlockingQueue<>( maxBatchSize );
+    }
+
+    @Override
+    public void start() throws Throwable
+    {
+        TransactionCommitProcess commitProcess = commitProcessSupplier.get();
+        txBatcher = new TransactionQueue( maxBatchSize, ( batch ) -> commitProcess.commit( batch, NULL, EXTERNAL ) );
+        lastQueuedTxId = lastAppliedTxId = txIdStoreSupplier.get().getLastCommittedTransactionId();
+    }
+
+    /**
+     * Queues a transaction for application.
+     *
+     * @param tx The transaction to be queued for application.
+     */
+    public void queue( CommittedTransactionRepresentation tx )
+    {
+        long receivedTxId = tx.getCommitEntry().getTxId();
+        long expectedId = lastQueuedTxId + 1;
+
+        if ( receivedTxId != expectedId )
+        {
+            log.warn( "Out of order txId: " + receivedTxId );
```

Comment:
Doesn't need to be there for the merge but we should log `expectedId` as well so we have more info on what's gone wrong.